apps/full-site-editing/full-site-editing-plugin/starter-page-templates/page-template-modal/index.js

```diff
@@ -29,46 +40,53 @@ class PageTemplateModal extends Component {
 	state = {
 		isLoading: false,
 		previewedTemplate: null,
-		blocksByTemplateSlug: {},
-		titlesByTemplateSlug: {},
 		error: null,
 		isOpen: false,
 	};
 
-	constructor( props ) {
-		super();
-		const hasTemplates = ! isEmpty( props.templates );
-		this.state.isOpen = hasTemplates;
-		if ( hasTemplates ) {
-			// Select the first template automatically.
-			this.state.previewedTemplate = this.getDefaultSelectedTemplate( props );
-			// Extract titles for faster lookup.
-			this.state.titlesByTemplateSlug = mapValues( keyBy( props.templates, 'slug' ), 'title' );
-		}
-	}
-
-	componentDidMount() {
-		if ( this.state.isOpen ) {
-			trackView( this.props.segment.id, this.props.vertical.id );
-		}
+	// Extract titles for faster lookup.
+	getTitlesByTemplateSlugs = memoize( templates =>
+		mapValues( keyBy( templates, 'slug' ), 'title' )
+	);
 
-		// Parse templates blocks and store them into the state.
-		const blocksByTemplateSlug = reduce(
-			this.props.templates,
+	// Parse templates blocks and store them into the state.
+	getBlocksByTemplateSlugs = memoize( templates =>
+		reduce(
+			templates,
 			( prev, { slug, content } ) => {
 				prev[ slug ] = content
 					? parseBlocks( replacePlaceholders( content, this.props.siteInformation ) )
 					: [];
 				return prev;
 			},
 			{}
-		);
+		)
+	);
+
+	static getDerivedStateFromProps( props, state ) {
+		// The only time `state.previewedTemplate` isn't set is before `templates`
+		// are loaded. As soon as we have our `templates`, we set it using
+		// `this.getDefaultSelectedTemplate`. Afterwards, the user can select a
+		// different template, but can never un-select it.
+		// This makes it a reliable indicator for whether the modal has just been launched.
+		if ( ! state.previewedTemplate && ! isEmpty( props.templates ) ) {
```

Comment:
We should take into consideration this will happen immediately at mount when `props.templates` in not `undefined`.

> getDerivedStateFromProps is invoked right before calling the render method, both on the initial mount and on subsequent updates